### ISSUE TITLE
Implementing into an existing app - styled.withTheme.withConfig is not a function

### ISSUE THREAD

**Issue by mattimck** [GetStream/stream-chat-react-native#181]
We are having some difficulty implementing this into an existing (rather messy) ReactApp. The error we receive whenever `stream-chat-react-native` is imported (even if we don't put the components in the view, just import the package at all), is this:

```
styled.withTheme.withConfig is not a function. (In 'styled.withTheme.withConfig({
  displayName: "dist__MessageTextContainer",
  componentId: "lvznht-0"
})', 'styled.withTheme.withConfig' is undefined)
```

We are confident it's something in our app, as it works fine with the same version of React, and Stream (and `styled-components`) running in a fresh app. We have gone as far as putting the example code directly into App.js in replacement of our own, just to test if it ever works, but there is definitely something about the massive combination of packages this app has which is conflicting. We've tried taking them out one by one, but that process is very difficult since so many of the packages have links in the native parts of the app.

One thing that were able to do to stop the error, was downgrade to `stream-chat-react-native` `v0.3.12`, and the error goes away (there are other problems, but it improves at least), and it appears again as of `0.4.0`, so it given the changes in that release, we are fairly confident it is indeed something to do with `MessageTextContainer`.

Our package dependencies are:

```
"@delightfulstudio/react-native-wheel-picker-android": "^2.1.3",
"@react-native-community/async-storage": "^1.3.3",
"@react-native-community/netinfo": "^3.2.1",
"axios": "^0.18.0",
"camelcase-keys": "^5.2.0",
"cocoa-lumberjack": "^3.0.5",
"country-list": "^2.1.0",
"dotenv-parse-variables": "^0.2.0",
"formik": "^1.5.1",
"immer": "^2.1.4",
"moment": "^2.24.0",
"native-base": "^2.12.2",
"node-fetch": "^2.3.0",
"prop-types": "^15.7.2",
"qs": "^6.7.0",
"react": "16.8.3",
"react-art": "^16.8.5",
"react-native": "0.59.9",
"react-native-android-open-settings": "^1.3.0",
"react-native-background-geolocation": "^3.2.2",
"react-native-camera": "pre-react-native-0.60.x",
"react-native-code-push": "^5.6.0",
"react-native-config": "^0.11.7",
"react-native-dash": "0.0.9",
"react-native-device-info": "2.1.2",
"react-native-document-picker": "^3.3.3",
"react-native-dropdownalert": "^3.9.2",
"react-native-elements": "^1.1.0",
"react-native-emoji": "^1.6.0",
"react-native-firebase": "^5.3.1",
"react-native-formik": "^1.7.5",
"react-native-gesture-handler": "^1.1.0",
"react-native-htmlview": "^0.13.0",
"react-native-image-picker": "^0.28.1",
"react-native-keyboard-aware-scroll-view": "^0.8.0",
"react-native-maps": "^0.24.2",
"react-native-modal-datetime-picker": "^6.1.0",
"react-native-normalize": "^1.0.1",
"react-native-notifications": "^2.0",
"react-native-snackbar": "^1.0.0",
"react-native-splash-screen": "^3.2.0",
"react-native-svg": "^9.3.7",
"react-native-vector-icons": "^6.4.2",
"react-native-video": "^4.4.0",
"react-native-web": "^0.11.2",
"react-native-webview": "^5.8.1",
"react-navigation": "^3.5.1",
"react-navigation-stack": "^1.1.1",
"react-redux": "^6.0.1",
"recompose": "^0.30.0",
"redux": "^4.0.1",
"redux-persist": "^5.10.0",
"redux-saga": "^1.0.2",
"redux-thunk": "^2.3.0",
"remote-redux-devtools": "^0.5.16",
"snakecase-keys": "^2.1.0",
"stream-chat": "^1.7.0",
"stream-chat-react-native": "^0.6.6",
"styled-components": "^4.2.0",
"uuid": "^3.3.2",
"yup": "^0.27.0"
```

Additionally have tried newer and older version of `stream-chat-react-native` but with no luck. At first we were convinced that the issue was a conflict between versions of `styled-components`, but we've tried matching versions, and also giving our existing version a different alias to prevent overlap, but that didn't help either. 

Again, we are clear it's something in our app, but we are at a loss to which of our packages is causing it (or if it's something else?).

Any help is extremely appreciated!

Cheers,
Matt

**Comment by vishalnarkhede** [comment]
Hey @mattimck I will check and get back to you on this one :)

**Comment by vishalnarkhede** [comment]
@mattimck Can you please give me following information (which may give me some more clue for debugging this issue)?

1. Your rollup config
2. Babel config
3. devDependencies

**Comment by mattimck** [comment]
Thanks for the quick reply @vishalnarkhede! I'm not sure we have a rollup config, but here are the other parts you requested:

Babel config:
```
module.exports = {
  presets: ['module:metro-react-native-babel-preset'],
  plugins: [
    "transform-inline-environment-variables",
    "babel-plugin-styled-components",
    ["@babel/plugin-proposal-decorators", {"legacy": true}],
    ["module-resolver", {
      "alias": {
        "~": "./src"
      }
    }]
  ],
  "sourceMap": "inline",
  "retainLines": true,
};
```

devDependencies:
```
  "devDependencies": {
    "@babel/core": "^7.4.0",
    "@babel/plugin-proposal-decorators": "^7.4.4",
    "@babel/runtime": "^7.4.2",
    "@react-native-community/eslint-config": "^0.0.3",
    "babel-core": "^7.0.0-bridge.0",
    "babel-eslint": "^10.0.1",
    "babel-jest": "^24.5.0",
    "babel-plugin-module-resolver": "^3.2.0",
    "babel-plugin-styled-components": "^1.10.0",
    "babel-plugin-transform-inline-environment-variables": "^0.4.3",
    "eslint": "^5.15.3",
    "eslint-config-airbnb": "^17.1.0",
    "eslint-config-prettier": "^4.1.0",
    "eslint-import-resolver-babel-module": "^5.1.0",
    "eslint-plugin-import": "^2.17.2",
    "eslint-plugin-jest": "^22.4.1",
    "eslint-plugin-jsx-a11y": "^6.2.1",
    "eslint-plugin-prettier": "^3.0.1",
    "eslint-plugin-react": "^7.12.4",
    "jest": "^24.5.0",
    "jest-fetch-mock": "^2.1.1",
    "jsdom": "^13.2.0",
    "metro-react-native-babel-preset": "^0.53.1",
    "prettier": "^1.16.4",
    "react-dom": "^16.8.1",
    "react-native-mock-render": "^0.1.2",
    "react-native-testing-library": "^1.7.0",
    "react-test-renderer": "16.8.3",
    "redux-devtools-extension": "^2.13.8",
    "redux-saga-tester": "^1.0.439"
  }
```

**Comment by mattimck** [comment]
Hi @vishalnarkhede vishalnarkhede, I was able to have another go at this today, and from what I can see, it is very similar to the issue at https://github.com/styled-components/styled-components/issues/2681. 

I was able to get things running well in app, but I needed to import `withTheme` differently in  `MessageTextContainer.js`. 

As per https://github.com/styled-components/styled-components/issues/2681#issuecomment-602493753, I can see that `import styled, { withTheme } from '@stream-io/styled-components';` in `MessageTextContainer.js` doesn't appear to work. 

However if I edit withTheme to import using something such as `const withTheme = require('@stream-io/styled-components').withTheme` it works 🎉. I have been unable to get this to install via npm, from a forked repo, so I think i'm still doing something wrong there (I keep getting an error like `native handler was not registered you should import stream-chat-expo`).

To prove it works in my app though, I was able to edit the build in `dist/index.js` directly, and I added an additional var below the require of `styled-components`:
```
var styled = require('@stream-io/styled-components');
var withTheme = styled.withTheme;
```
and then changed:
```
var MessageTextContainer = styled.withTheme(function (props) {
```
to:
```
var MessageTextContainer = withTheme(function (props) {
```

This made everything work fine for me. Either way, both of those options are a hack, and not ideal. If anyone has any suggestions of how else I could resolve this, i'd really appreciate the help! Thank you!

**Comment by vishalnarkhede** [comment]
Hey @mattimck thanks for investigation. I understand the issue now ... let me see what I can do about it.

**Comment by mattimck** [comment]
Sorry @vishalnarkhede, i'm at a loss why it makes any difference either. But I've switched just those lines back and forward over and over and without fail it fixes it/breaks it each time. 🤷‍♂️ 

I don't fully understand how the rollup output works exactly, so I'm having trouble getting any further, and unfortunately it doesn't look like the commenter in https://github.com/styled-components/styled-components/issues/2681#issuecomment-602493753 has fully understood why it's happening (or something similar, at least) either.

I'd be happy to try any other suggestions you might have - i'm really keen to get this working if possible, even if it requires us putting some further work into our code / stream-chat code or styled-components code, as with this "hack" in place stream chat works great in our app!

Thank you again!

**Comment by vishalnarkhede** [comment]
@mattimck I will do more investigation and get back to you!

**Comment by vishalnarkhede** [comment]
@mattimck I think I found the issue. Can you add following alias in your babel config:

```
"@stream-io/styled-components": "@stream-io/styled-components/native/dist/styled-components.native.cjs.js"
```

So basically your full babel config will be as following:

```
module.exports = {
  presets: ['module:metro-react-native-babel-preset'],
  plugins: [
    "transform-inline-environment-variables",
    "babel-plugin-styled-components",
    ["@babel/plugin-proposal-decorators", {"legacy": true}],
    ["module-resolver", {
      "alias": {
        "~": "./src",
        '@stream-io/styled-components':
          '@stream-io/styled-components/native/dist/styled-components.native.cjs.js',
      }
    }]
  ],
  "sourceMap": "inline",
  "retainLines": true,
};
```

Don't forget to reset cache - `react-native start --resetCache` or `yarn start --reset-cache` for updated babel config to work :)

Let me know if it works!!

**Comment by vishalnarkhede** [comment]
@mattimck ^^

**Comment by mattimck** [comment]
Hi @vishalnarkhede - sorry for slow reply, i've been unwell the last couple of days! I have tried this quickly this morning, and unfortunately it doesn't appear to resolve the proble - however I will try again later today and see if I can get any further progress with more time applied to it. Thank you again for your help!

**Comment by vishalnarkhede** [comment]
@mattimck 
So basically what my suspicion is, when your webpack (assuming you are using webpack) resolves modules, its using `@stream-io/styled-components/dist/styled-components.cjs.js` instead of `@stream-io/styled-components/native/dist/styled-components.native.cjs.js`. 

Can you please confirm if thats not the case, by adding some `throw Error('not using correct package')` in `@stream-io/styled-components/dist/styled-components.cjs.js`. Something like following:

```js
var styled = function styled(tag) {
  throw Error('not using correct package')
  return constructWithOptions(createStyledComponent, tag);
};
```

**Comment by mattimck** [comment]
@vishalnarkhede - thanks, I tried this, but it did _not_ throw the error. 

Just to be sure though, I did try to add the same code into `@stream-io/styled-components/native/dist/styled-components.native.cjs.js` and it _did_ throw the error, so i'm pretty sure it's using the correct one?

I also tried changing my babel config to deliberately use `@stream-io/styled-components/dist/styled-components.cjs.js` just to be sure i'd set the `throw Error` correctly, but it complained immediately: `Unhandled JS Exception: styled__default.TextInput is not a function.`. Either way, I think shows that the alias is working correctly in the babel config, right?
